client.go

```diff
@@ -0,0 +1,161 @@
+// Package checkout contains Go client for Checkout API.
+//
+// Checkout API documentation: https://docs.checkout.com/v2.0
+package checkout
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"io"
+	"io/ioutil"
+	"net/http"
+
+	"github.com/pkg/errors"
+)
+
+// Caller makes HTTP call with given options and decode response into given struct.
+// Client implements this interface and pass itself to entity clients. You may create entity clients with own caller for
+// test purposes.
+type Caller interface {
+	Call(ctx context.Context, method, path string, idempotencyKey string, reqObj interface{}, respObj interface{}) (statusCode int, callErr error)
+}
+
+// HTTPClient is interface fot HTTP client. Built-in net/http.Client implements this interface as well.
+type HTTPClient interface {
+	Do(r *http.Request) (*http.Response, error)
+}
+
+// Option is a callback for redefine client parameters.
+type Option func(*Client)
+
+// Client contains API parameters and provides set of API entity clients.
+type Client struct {
+	httpClient HTTPClient
+	endpoint   string
+	secretKey  string
+}
+
+const (
+	EndpointLive    = "https://api.checkout.com"
+	EndpointSandbox = "https://api.sandbox.checkout.com"
+
+	headerAuthorization = "Authorization"
+	headerIdempotency   = "Cko-Idempotency-Key"
+)
+
+// New creates new client with given options.
+func New(options ...Option) *Client {
+	c := &Client{
+		httpClient: http.DefaultClient,
+		endpoint:   EndpointLive,
+	}
+
+	for _, option := range options {
+		option(c)
+	}
+
+	return c
+}
+
+// OptHTTPClient returns option with given HTTP client.
+func OptHTTPClient(httpClient HTTPClient) Option {
+	return func(c *Client) {
+		c.httpClient = httpClient
+	}
+}
+
+// OptSecretKey returns option with given secret key.
+func OptSecretKey(secretKey string) Option {
+	return func(c *Client) {
+		c.secretKey = secretKey
+	}
+}
+
+// OptEndpoint returns option with given API endpoint.
+func OptEndpoint(endpoint string) Option {
+	return func(c *Client) {
+		c.endpoint = endpoint
+	}
+}
+
+// Call does HTTP request with given params using set HTTP client. Response will be decoded into respObj.
+// ServerError may be returned if something went wrong. If API return error as response, then Call returns error of type checkout.ServerError.
+func (c *Client) Call(ctx context.Context, method, path string, idempotencyKey string, reqObj interface{}, respObj interface{}) (statusCode int, callErr error) {
```

Comment:
`PaymentClient` should handle other codes by itself to detect business errors.

Reply:
PaymentClient handles error returned from the Call method anyways and it can handle other codes by casting error to ServerError. It will result in a cleaner code here and there I believe.

Reply:
`ServerError` contains decoded JSON also. Only base client knows hot to decode it right.

Reply:
You'll have to rewrite this code as soon as there will be some error which body you will need to process in the Payment/Capture/Whatever clients. It's still looks fragile to me.

Reply:
I can remove `ServerError.StatusCode` to remove status code duplication, if you wish. But I added it to `ServerError` just to see verbose error in logs.

Reply:
Ok, let's leave it as is and wait